File: integrations/paycom/preferences.ts
import { z } from 'zod';
import type { Storage } from '../../services/storage/index.js';
import type { Audit } from '../../services/audit/index.js';
import { assert } from '../../shared/errors.js';
import {
  paycomDefaults,
  type PaycomPreferences,
  type PaycomSettings,
} from '../../shared/paycom.js';
export const preferencesSchema = z
  .object({
    automatic_sync: z.boolean(),
    sync_interval_seconds: z.union([
      z.literal(1800),
      z.literal(3600),
      z.literal(7200),
      z.literal(14400),
    ]),
    opening_page: z.enum(['timecards', 'employees']),
    rows_per_page: z.union([z.literal(25), z.literal(50), z.literal(100)]),
    name_order: z.enum(['first_last', 'last_first']),
    default_sort: z.enum(['employeeName', 'condition', 'inDay']),
    department: z.string().max(200).nullable(),
    station: z.string().max(200).nullable(),
    columns: z
      .array(z.enum(['inDay', 'outLunch', 'inLunch', 'outDay', 'totalHours', 'condition']))
      .max(6)
      .refine((v) => new Set(v).size === v.length),
    driver_departments: z.array(z.string().max(200)).max(500).nullable(),
  })
  .strict();
type Stored = Pick<PaycomSettings, 'revision' | 'values' | 'history'>;
export function readPaycomSettings(storage: Storage, dspId: string): PaycomSettings {
  return storage.dsp(dspId, (db) => {
    const raw = db.one<{ value: string }>(
      "SELECT value FROM settings WHERE key='paycom.preferences'",
    );
    const stored: Stored = raw
      ? JSON.parse(raw.value)
      : { revision: 0, values: paycomDefaults, history: [] };
    const schedule = db.one<{ enabled: number }>(
      "SELECT enabled FROM schedules WHERE provider='paycom'",
    )!;
    return {
      ...stored,
      values: { ...stored.values, automatic_sync: Boolean(schedule.enabled) },
      options: {
        departments: db.all<{ value: string; count: number }>(
          'SELECT department value,count(*) count FROM employees WHERE publication_id=(SELECT id FROM publications WHERE active=1) GROUP BY department ORDER BY department',
        ),
        stations: db
          .all<{
            station: string;
          }>(
            'SELECT DISTINCT station FROM employees WHERE publication_id=(SELECT id FROM publications WHERE active=1) ORDER BY station',
          )
          .map((v) => v.station),
      },
    };
  });
}
export function writePaycomSettings(
  storage: Storage,
  audit: Audit,
  dspId: string,
  actorId: string,
  revision: number,
  values: PaycomPreferences,
) {
  preferencesSchema.parse(values);
  const before = readPaycomSettings(storage, dspId);
  storage.dsp(dspId, (db) =>
    db.transaction(() => {
      const raw = db.one<{ value: string }>(
        "SELECT value FROM settings WHERE key='paycom.preferences'",
      );
      const current = raw ? JSON.parse(raw.value).revision : 0;
      assert(current === revision, 'settings_changed_reload_before_saving', 409);
      const connected = db.one<{ enabled: number }>(
        "SELECT enabled FROM connections WHERE provider='paycom'",
      )!;
      assert(
        !values.automatic_sync || connected.enabled,
        'connect_paycom_before_automatic_sync',
        409,
      );
      const stored: Stored = {
        revision: revision + 1,
        values,
        history: [
          { revision: before.revision, at: new Date().toISOString(), values: before.values },
          ...before.history,
        ].slice(0, 20),
      };
      db.run(
        "INSERT INTO settings(key,value) VALUES ('paycom.preferences',?) ON CONFLICT(key) DO UPDATE SET value=excluded.value",
        JSON.stringify(stored),
      );
      const scheduleChanged =
        !db.one("SELECT 1 FROM settings WHERE key='paycom.syncIntervalSeconds'") ||
        before.values.automatic_sync !== values.automatic_sync ||
        before.values.sync_interval_seconds !== values.sync_interval_seconds;
      db.run(
        "INSERT INTO settings(key,value) VALUES ('paycom.syncIntervalSeconds',?) ON CONFLICT(key) DO UPDATE SET value=excluded.value",
        String(values.sync_interval_seconds),
      );
      if (scheduleChanged)
        db.run(
          "UPDATE schedules SET enabled=?,next_run=? WHERE provider='paycom'",
          Number(values.automatic_sync),
          values.automatic_sync
            ? new Date(Date.now() + values.sync_interval_seconds * 1000).toISOString()
            : null,
        );
    }),
  );
  audit.record(actorId, dspId, 'paycom.settings_updated', `Revision ${revision + 1}`);
  return readPaycomSettings(storage, dspId);
}

File: backend/src/workforce/preferences.rs
use crate::{
    Error, Result,
    collectors::Provider,
    db::{AuditChange, Db, Store, iso, n},
    ensure, validate as v,
};
use serde_json::{Value, json};
use std::collections::HashSet;
const COLUMNS: [&str; 6] = [
    "inDay",
    "outLunch",
    "inLunch",
    "outDay",
    "totalHours",
    "condition",
];
pub fn defaults() -> Value {
    json!({"opening_page":"timecards","rows_per_page":100,"name_order":"first_last",
        "default_sort":"employeeName","department":null,"station":null,"columns":COLUMNS,
        "driver_departments":null,"late_da_time":"10:01","late_da_departments":[]})
}
pub(super) fn preferences(db: &Db) -> Result<Value> {
    let mut stored = db.setting(
        "paycom.preferences",
        json!({"revision":0,"values":defaults(),"history":[]}),
    )?;
    // Preferences saved before a key existed take that key's default.
    for (key, value) in defaults().as_object().unwrap() {
        if stored["values"].get(key).is_none() {
            stored["values"][key] = value.clone();
        }
    }
    without_retired(&mut stored["values"]);
    Ok(stored)
}
// Collection schedules replaced these. Preferences saved through v0.0.9 hold
// them, and a dashboard opened before an update still sends them.
fn without_retired(values: &mut Value) {
    if let Some(values) = values.as_object_mut() {
        values
            .retain(|key, _| !["automatic_sync", "sync_interval_seconds"].contains(&key.as_str()));
    }
}
fn validate_preferences(value: &Value) -> Result<()> {
    v::fields(
        value,
        &[
            "opening_page",
            "rows_per_page",
            "name_order",
            "default_sort",
            "department",
            "station",
            "columns",
            "driver_departments",
            "late_da_time",
            "late_da_departments",
        ],
    )?;
    v::choice(
        value,
        "opening_page",
        &["timecards", "meal-breaks", "employees"],
    )?;
    v::choice(value, "name_order", &["first_last", "last_first"])?;
    v::choice(
        value,
        "default_sort",
        &["employeeName", "condition", "inDay"],
    )?;
    ensure(
        [25, 50, 100].contains(&v::integer(value, "rows_per_page", 1, 100)?),
        "invalid_input",
        400,
    )?;
    for key in ["department", "station"] {
        if !value[key].is_null() {
            v::text(value, key, 0, 200)?;
        }
    }
    let cols = value["columns"]
        .as_array()
        .ok_or_else(|| Error::new("invalid_input", 400))?;
    ensure(
        cols.len() <= 6
            && cols
                .iter()
                .all(|c| c.as_str().is_some_and(|s| COLUMNS.contains(&s)))
            && cols
                .iter()
                .map(Value::to_string)
                .collect::<HashSet<_>>()
                .len()
                == cols.len(),
        "invalid_input",
        400,
    )?;
    for key in ["driver_departments", "late_da_departments"] {
        // Only the Timecard filter distinguishes "all" (null) from "none" (empty).
        if key == "driver_departments" && value[key].is_null() {
            continue;
        }
        ensure(
            value[key].as_array().is_some_and(|a| {
                a.len() <= 500
                    && a.iter()
                        .all(|x| x.as_str().is_some_and(|s| s.chars().count() <= 200))
            }),
            "invalid_input",
            400,
        )?;
    }
    let time = v::text(value, "late_da_time", 5, 5)?.as_bytes();
    ensure(
        time[2] == b':'
            && [0, 1, 3, 4].iter().all(|i| time[*i].is_ascii_digit())
            && (time[0], time[1]) <= (b'2', b'3')
            && time[3] <= b'5',
        "invalid_input",
        400,
    )?;
    Ok(())
}
// The settings a member can change, compared for the audit log. An unset
// filter means "All", and an empty list "None".
fn preference_changes(before: &Value, after: &Value) -> Vec<AuditChange> {
    const FIELDS: [(&str, &str); 10] = [
        ("opening_page", "paycom.opening_page"),
        ("rows_per_page", "paycom.rows_per_page"),
        ("name_order", "paycom.name_order"),
        ("default_sort", "paycom.default_sort"),
        ("department", "paycom.department"),
        ("station", "paycom.station"),
        ("columns", "paycom.columns"),
        ("driver_departments", "paycom.driver_departments"),
        ("late_da_time", "paycom.late_da_time"),
        ("late_da_departments", "paycom.late_da_departments"),
    ];
    let text = |value: &Value| match value {
        Value::Null => "All".to_owned(),
        Value::String(text) => text.clone(),
        Value::Array(items) if items.is_empty() => "None".to_owned(),
        Value::Array(items) => items
            .iter()
            .map(|item| {
                item.as_str()
                    .map_or_else(|| item.to_string(), str::to_owned)
            })
            .collect::<Vec<_>>()
            .join(", "),
        other => other.to_string(),
    };
    FIELDS
        .iter()
        .filter(|(key, _)| before[key] != after[key])
        .map(|(key, field)| (*field, Some(text(&before[key])), Some(text(&after[key]))))
        .collect()
}
impl Store {
    pub fn preferences(&self, id: &str) -> Result<Value> {
        let db = self.collector(id, Provider::Paycom)?;
        let mut out = preferences(&db)?;
        let departments=db.all("SELECT department value,count(*) count FROM employees WHERE \
            publication_id=(SELECT id FROM publications WHERE active=1) GROUP BY department ORDER BY department",[])?;
        let stations: Vec<Value> = db
            .all(
                "SELECT DISTINCT station FROM employees WHERE \
            publication_id=(SELECT id FROM publications WHERE active=1) ORDER BY \
            station",
                [],
            )?
            .into_iter()
            .map(|r| r["station"].clone())
            .collect();
        out["options"] = json!({"departments":departments,"stations":stations});
        Ok(out)
    }
    pub fn save_preferences(
        &self,
        id: &str,
        actor: &str,
        revision: i64,
        values: &Value,
    ) -> Result<Value> {
        let mut values = values.clone();
        without_retired(&mut values);
        let values = &values;
        validate_preferences(values)?;
        let previous = self.preferences(id)?;
        let db = self.collector(id, Provider::Paycom)?;
        db.transaction(|| {
            let before = preferences(&db)?;
            ensure(
                n(&before, "revision") == revision,
                "settings_changed_reload_before_saving",
                409,
            )?;
            let mut history =
                vec![json!({"revision":before["revision"],"at":iso(),"values":before["values"]})];
            history.extend(before["history"].as_array().cloned().unwrap_or_default());
            history.truncate(20);
            db.set(
                "paycom.preferences",
                &json!({"revision":revision+1,"values":values,"history":history}),
            )
        })?;
        self.audit_with(
            Some(actor),
            Some(id),
            "paycom.settings_updated",
            &format!("Revision {}", revision + 1),
            None,
            &preference_changes(&previous["values"], values),
        )?;
        self.preferences(id)
    }
}
